FROST for Spend Authorization Signatures (#3)
ZcashFoundation/zips: zip-frost.rst

::

  ZIP: Unassigned {numbers are assigned by ZIP editors}
  Title: FROST for Spend Authorization Signatures

>   Owners: Chelsea Komlo <ckomlo@uwaterloo.ca>
>           Conrado Gouvea <conrado@zfnd.org>
>           Deirdre Connolly <deirdre@zfnd.org>

Reviewer: ```suggestion
  Owners: Conrado Gouvea <conrado@zfnd.org>
          Chelsea Komlo <ckomlo@uwaterloo.ca>
          Deirdre Connolly <deirdre@zfnd.org>
```

  Credits: First Credited
           ...
  Status: Draft
  Category: Wallet
  Created: 2022-08-dd
  License: {usually MIT}
  Pull-Request: <https://github.com/zcash/zips/pull/TODO>

Terminology
===========

{Edit this to reflect the key words that are actually used.}
The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to
be interpreted as described in RFC 2119. [#RFC2119]_

The terms below are to be interpreted as follows:

{Term to be defined}
  {Definition.}
{Another term}
  {Definition.}

Abstract
========

> This proposal defines the usage of FROST [#FROST]_, a threshold signature scheme,

Reviewer 2: I would say that this is about adapting FROST to be unlinkable, which is a requirement for its use in the Zcash protocol.

Author: Added this wording in cf19f23

to generate signatures compatible with spend authorization signatures in the Zcash
protocol, for the Sapling and Orchard network upgrades.

Motivation
==========

In the Zcash protocol, Spend Authorization Signatures are employed to effectively
authorize a transaction. The ability to generate these signatures with the user's
private key is what effectively allows the to spend funds.

This is security-critical step, since anyone who obtains access to the private
key will be able to spend the user's funds. For this reason, one interesting
possibility is to require multiple entities to allow the transaction to go through.
This can be accomplished with threshold signatures, where... (TODO: explain)

> FROST is one of such threshold signatures protocols. (TODO: more about FROST)

Reviewer 2: However, FROST as-is cannot be used, since the Zcash protocol also requires re-randomizing public and private keys to ensure unlinkability between transactions.

Author: Added in cf19f23

Requirements
============

- All signatures generated by following this ZIP MUST be verified successfully
  as Sapling or Orchard spend authorization signatures.
- (Must keep same security requirements from the spec?)
- (Randomizer security requirements?)

> 

Reviewer: ```suggestion
- The party that generates alpha should also be the party that has access to the witness for the proof.
```

> Non-requirements

Reviewer 2: I would also add a section about the threat model, and how this assumes a trusted prover that is also the coordinator. In the future, we can write a different ZIP for a less-trusted coordinator.

Author: Done in cf19f23

================

- (Talk about how proof generation is not covered?)

Specification
=============

Re-randomized FROST
-------------------

To add re-randomization to FROST, follow the specification [#FROST]_ with the
following modifications:

In Round Two, the Coordinator also generates a random scalar α by following the

> RedDSA.GenRandom procedure and sends it to each signer along with the message

Reviewer 2: Over a secure channel?

Reviewer 2: I would specify that function specifically- i.e, something like `generate_randomizer` ? And say that the coordinator performs this.

Author: Both done in cf19f23

and the set of signing commitments.

The `sign` function is changed to receive α and incorporate it into the
computation of `sig_share`. It is specified as the following: ::

  Inputs:
  - identifier, Identifier i of the signer. Note identifier will never equal 0.
  - sk_i, Signer secret key share, a Scalar.
  - group_public_key, public key corresponding to the group signing key,
    an Element in G.
  - nonce_i, pair of Scalar values (hiding_nonce, binding_nonce) generated in
    round one.
  - msg, the message to be signed (sent by the Coordinator).
  - commitment_list =
      [(j, hiding_nonce_commitment_j, binding_nonce_commitment_j), ...], a
    list of commitments issued in Round 1 by each signer and sent by the Coordinator.
    Each element in the list indicates the signer identifier j and their two commitment
    Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).s
    This list MUST be sorted in ascending order by signer identifier.
  - alpha, the randomizer Scalar.

  Outputs: a Scalar value representing the signature share

>   def sign(identifier, sk_i, group_public_key, nonce_i, msg, commitment_list, alpha):

Author: Everything is the same except the new `alpha` parameter...

    # Encode the commitment list
    encoded_commitments = encode_group_commitment_list(commitment_list)

    # Compute the binding factor

>     binding_factor = compute_binding_factor(encoded_commitments, msg)

Reviewer 3: Transcript alpha as well?

Author: Also working on that, thanks!

Author: Done in cf19f23

    # Compute the group commitment
    group_commitment = compute_group_commitment(commitment_list, binding_factor)

    # Compute Lagrange coefficient
    participant_list = participants_from_commitment_list(commitment_list)
    lambda_i = derive_lagrange_coefficient(identifier, participant_list)

    # Compute the per-message challenge
    challenge = compute_challenge(group_commitment, group_public_key, msg)

    # Compute the signature share
    (hiding_nonce, binding_nonce) = nonce_i

>     sig_share = hiding_nonce + (binding_nonce * binding_factor) + (lambda_i * (sk_i + alpha / len(commitment_list))  * challenge)

Author: And the addition of `alpha / len(commitment_list)` here

>     sig_share = hiding_nonce + (binding_nonce * binding_factor) + (lambda_i * (sk_i + alpha / len(commitment_list))  * challenge)

Reviewer 3: Is this formula correct? `sig_share` would need to be `hiding_nonce + (binding_nonce * binding_factor) + (lambda_i * sk_i + alpha / len(commitment_list)) * challenge)` if the intention is for this signature to verify as `K + alphaG`, as I believe given the coordinator's calculation of `alphaG`.

Author: Yep it's wrong, I'm working on fixing it. Thanks for spotting it!

Reviewer 3: Happy to. I was concerned I'd have to change my work significantly, yet seems like I won't.

While this is the model I took, as it creates a leaderless protocol, it is also possible for the coordinator to simply add alpha * challenge to the signature, no? It'd still be broadcasted though so everyone knows which key they're signing for (as the key is part of the challenge in general, though I'd have to double check RedDsa's policy here).

This would translate into a protocol without a coordinator by having every participant do so on their own. The only issue I had with this was regarding implementation. I found it beneficial to create constructions where the share signing function advances the state. If you have to sign for multiple shares, you cannot advance the state as such, assuming advancing the state once to X prevents advancing again to X. While you could sign for (sk_i + alpha) in the first place, offering a single sign, you then have to strip alpha * challenge out for the part communicated. It ends up very annoying to implement, so I do prefer the inversion based model present here. Regardless, I wanted to discuss the single-signer modification model as well, while I had the opportunity, and see what other peoples' thoughts were.

Author: Yeah in the version I'm working on, I'm making the coordinator add `alpha * challenge` as you said. I think it would still be leaderless since all participants now alpha so they can all add it to the final signature?

Sorry, what do you mean by state?

Author: Changed formula in cf19f23

Reviewer 3: To preface, my priority with FROST is modularity and composability, not Zcash, so bit of a different focus. I also want to work on consistency in the ecosystem though, which is why my work is fully usable in an IETF compliant fashion, and why I'd like to hear it can be used without issue under this ZIP as well.

I have a function, sign_share, in my trait which allows for an arbitrary algorithm to specify how to produce its share. The following is an excerpt implementing FROST over a ring signature scheme.

```
  fn sign_share(
    &mut self,
    view: &FrostView<Ed25519>,
    nonce_sum: dfg::EdwardsPoint,
    b: dfg::Scalar,
    nonce: dfg::Scalar,
    msg: &[u8]
  ) -> dfg::Scalar {
    // Apply the binding factor to the H variant of the nonce
    self.AH.0 += self.AH.1 * b;

    // Use the transcript to get a seeded random number generator
    // The transcript contains private data, preventing passive adversaries from recreating this
    // process even if they have access to commitments (specifically, the ring index being signed
    // for, along with the mask which should not only require knowing the shared keys yet also the
    // input commitment masks)
    let mut rng = ChaCha12Rng::from_seed(self.transcript.rng_seed(b"decoy_responses"));

    self.msg = Some(msg.try_into().expect("CLSAG message should be 32-bytes"));

    #[allow(non_snake_case)]
    let (clsag, pseudo_out, p, c) = Clsag::sign_core(
      &mut rng,
      &self.image,
      &self.input(),
      self.mask(),
      &self.msg.as_ref().unwrap(),
      nonce_sum.0,
      self.AH.0.0
    );
    self.interim = Some(Interim { p, c, clsag, pseudo_out });

    let share = dfg::Scalar(nonce.0 - (p * view.secret_share().0));

    share
  }
```

Compared to EdDSA/moden Schnorr:

```
  fn sign_share(
    &mut self,
    params: &FrostView<C>,
    nonce_sum: C::G,
    _: C::F,
    nonce: C::F,
    msg: &[u8],
  ) -> C::F {
    let c = H::hram(&nonce_sum, &params.group_key(), msg);
    self.c = Some(c);
    schnorr::sign::<C>(params.secret_share(), nonce, c).s
  }
```

There are two relevant distinctions.
1) The former, which expects to only be called once, performs a state mutation which is invalid if applied multiple times. The latter does not.
2) The former is `nonce.0 - (p * view.secret_share().0`. The latter resolves to `nonce + (private_key * challenge)`. For some reason, the classical Schnorr subtraction was used when calculating the solution.

Accordingly, this trait must be made contextual of signing involving an additive offset in order to properly do so. That is not the case with the linearly applied offset (nor an offset which only one party applies, which is technically with leader, yet given a proper protocol for generating the randomizer, ends up meaning one person does one additional scalar addition other parties do not, which is not of significance besides the lack of ability to claim "leaderless").

While I don't think this remains relevant to Zcash, and while I do believe the coordination/each participant simply adding this themselves is optimal in that regard, it does decrease the level at which abstractions can be performed. That's the comment against it, which I wanted to properly explain.

    return sig_share

> When the Coordinator outputs the signature, the Coordinator must also compute and

Reviewer 2: You'll also need to define a new verification function for the coordinator to perform, using the signers' public keys and alpha

Author: Good point, done in cf19f23

The terms below are to be interpreted as follows:

{Term to be defined}
  {Definition.}
{Another term}
  {Definition.}

Abstract
========

This proposal adapts FROST [#FROST]_, a threshold signature scheme,
to make it unlinkable, which is a requirement for its use in the Zcash protocol.
The adapted scheme generates signatures compatible with spend authorization
signatures in the Zcash protocol, for the Sapling and Orchard network upgrades.

Motivation
==========

In the Zcash protocol, Spend Authorization Signatures are employed to authorize
a transaction. The ability to generate these signatures with the user's
private key is what effectively allows the user to spend funds.

This is a security-critical step, since anyone who obtains access to the private
key will be able to spend the user's funds. For this reason, one interesting
possibility is to require multiple users to allow the transaction to go through.
This can be accomplished with threshold signatures, where the private key is
split between users in a way that a threshold (e.g. 2 out of 3) of them must
sign the transaction in order to create the final signature. This enables scenarios

> such as users and exchanges sharing custody of a wallet, for example.

Author: Mention that a user can be something other than a person (e.g. a device)

Author: Done by using "parties" in 80a22f4

> FROST is one of such threshold signatures protocols. However, it can't be used as-is

Author: ```suggestion
FROST is one of such threshold signature protocols. However, it can't be used as-is
```

since the Zcash protocol also requires re-randomizing public and private keys
to ensure unlinkability between transactions. This ZIP specifies a variant of
FROST with re-randomization support.

Requirements
============

> - All signatures generated by following this ZIP MUST be verified successfully

Author: ```suggestion
- All signatures generated by following this ZIP must be verified successfully.
```

>   as Sapling or Orchard spend authorization signatures.

Author: MUST be successfully verifiable

- The signatures generated by following this ZIP should meet the security criteria
  for Signature with Re-Randomizable Keys as specified in the Zcash protocol [#protocol-concretereddsa]_.

> - Follow the threat model described below.

Author: ```suggestion
- The threat model described below must be taken into account.
```

Threat Model
------------

In normal usage, a Zcash user follows multiple steps in order to generate a
shielded transaction:

- The transaction is created.

> - The transaction is signed with a randomized version of the user's spend

Author: ```suggestion
- The transaction is signed with a re-randomized version of the user's spend
```

(also any other instances)

  authorization private key.
- The zero-knowledge proof for the transaction is created with the randomizer
  as an auxiliary (secret) input, among others.

When employing re-randomized FROST as specified in this ZIP, the goal is to

> split the spend authorization private key among multiple participants. This means

Author: ```suggestion
split the spend authorization private key :math:`\mathsf{ask}` among multiple possible signers. This means
```

that the proof generation will still be generated by a single participant, likely
the one that created the transaction in the first place and that will actually

> submit the transaction to the network. Note that his user already controls the

Author: ```suggestion
the one that created the transaction in the first place. Note that this user already controls the
```

privacy of the transaction since they are responsible for creating the proof.

> This fits well into the "Coordinator" role from the FROST specification. The

Author: add reference

Author: done in 80a22f4

Coordinator is responsible for sending the message to be signed to all participants,
and to aggregate the signature shares.

With those considerations in mind, the threat model considered in this ZIP is:

- The Coordinator is trusted with the privacy of the transaction (which includes

>   the unlinkability property). A rogue Coordinator will be able to break

Author: add definition of unlinkability in terminology section

Author: Done in 80a22f4, but unsure about the definition

  unlinkability and privacy, but should not be able to create signed transactions
  without the approval of `MIN_SIGNERS` participants, as specified in FROST.
- All key share holders are also trusted with the privacy of the transaction.
  A future specification may support a scenario where individual key share
  holders are not trusted with it.

Non-requirements

> ================

Author: Mention key generation here

Author: Also network privacy

Author: Done in 80a22f4

> - This ZIP does not support removing the Coordinator role, as described in #[FROST]_.

Author: refer to section in RFC

Author: done in 80a22f4

> - This ZIP does not support removing the Coordinator role, as described in #[FROST]_.

Reviewer: https://github.com/cfrg/draft-irtf-cfrg-frost/blob/master/draft-irtf-cfrg-frost.md#removing-the-coordinator-role-no-coordinator

- This ZIP does not prevent key share holders from breaking the privacy

>   of the transaction.

Author: ```suggestion
- This ZIP does not prevent key share holders from linking the signing operation to a transaction in the blockchain.
```

Specification
=============

Re-randomized FROST
-------------------

To add re-randomization to FROST, follow the specification [#FROST]_ with the
following modifications.

Randomizer Generation
'''''''''''''''''''''

> A new helper function is defined, which computes RedDSA.GenRandom:

Author: ```suggestion
A new helper function is defined, which computes :math:`\mathsf{RedDSA.GenRandom}`:
```

> ::

Author: Mention that it's python pseudo-code

Author: Done in 80a22f4

  randomizer_generate():

  Inputs:
  - None

>   Outputs: randomizer, a Scalar

Author: refer to RFC for notation

Author: done in 80a22f4

  def randomizer_generate():
    randomizer_input = random_bytes(64)
    return H2(randomizer_input)

Binding Factor Computation
''''''''''''''''''''''''''

The `compute_binding_factor` function is changed to receive the `randomizer`
as follows: ::

  Inputs:
  - encoded_commitment_list, an encoded commitment list (as computed
    by encode_group_commitment_list)
  - msg, the message to be signed.
  - randomizer, the randomizer Scalar.

  Outputs: A Scalar representing the binding factor

  def compute_binding_factor(encoded_commitment_list, msg, randomizer):
    msg_hash = H3(msg)
    rho_input = encoded_commitment_list || msg_hash || G.SerializeScalar(randomizer)
    binding_factor = H1(rho_input)
    return binding_factor

Round Two - Signature Share Generation
''''''''''''''''''''''''''''''''''''''

> In Round Two, the Coordinator generates a random scalar `randomizer` by calling

Author: mention the scalar generated above

Author: Above, just the scalar generation procedure was defined, but it was not generated yet.

`randomizer_generate` and sends it to each signer, over a secret channel,
along with the message and the set of signing commitments. (Note that this differs
from regular FROST which just requires an authenticated channel.)

The `sign` function is changed to receive `randomizer` and incorporate it into the
computation of the binding factor. It is specified as the following: ::

  Inputs:

>   - identifier, Identifier i of the signer. Note identifier will never equal 0.

Reviewer 4: Make it clearer that this is not a "note identifier":

```suggestion
  - identifier, Identifier i of the signer.
    Note: identifier will never equal 0.
```

    # Compute the randomized group public key
    randomized_group_public_key = group_public_key + G * randomizer

    # Encode the commitment list
    encoded_commitments = encode_group_commitment_list(commitment_list)

    # Compute the binding factor
    binding_factor = compute_binding_factor(encoded_commitments, msg, randomizer)

    # Compute the group commitment
    group_commitment = compute_group_commitment(commitment_list, binding_factor)

    # Compute Lagrange coefficient
    participant_list = participants_from_commitment_list(commitment_list)
    lambda_i = derive_lagrange_coefficient(identifier, participant_list)

    # Compute the per-message challenge
    challenge = compute_challenge(group_commitment, randomized_group_public_key, msg)

    # Compute the signature share
    (hiding_nonce, binding_nonce) = nonce_i
    sig_share = hiding_nonce + (binding_nonce * binding_factor) + (lambda_i * sk_i * challenge)

    return sig_share

Signature Share Verification and Aggregation
''''''''''''''''''''''''''''''''''''''''''''

The `verify_signature_share` is changed to incorporate the randomizer,
as follows: ::

  Inputs:

>   - identifier, Identifier i of the signer. Note identifier will never equal 0.

Reviewer 4: ```suggestion
  - identifier, Identifier i of the signer.
    Note: identifier will never equal 0.
```

Ciphersuites
------------

FROST(Jubjub, BLAKE2b-512)
'''''''''''''''''''''''''''''

This ciphersuite uses Jubjub for the Group and BLAKE2b-512 for the Hash function `H`
meant to produce signatures indistinguishable from RedJubjub Sapling Spend
Authorization Signatures as specified in [#protocol-concretespendauthsig]_.

- Group: Jubjub [#protocol-jubjub]_

  - Order: 6554484396890773809930967563523245729705921265872317281365359162392183254199 (see [#protocol-jubjub]_)
  - Identity: as defined in [#protocol-jubjub]_
  - RandomScalar: Implemented by generating a random 64-byte string and invoking
    DeserializeScalar on the result
  - RandomNonZeroScalar: Implemented by generating a random 32-byte string that
    is not equal to the all-zero string and invoking DeserializeScalar on the result.
  - SerializeElement(P): Implemented as :math:`\mathsf{repr}_\mathbb{J}(P)` as defined in [#protocol-jubjub]_
  - DeserializeElement(P): Implemented as :math:`\mathsf{abst}_\mathbb{J}(P)` as defined in [#protocol-jubjub]_
  - SerializeScalar: Implemented by outputting the little-endian 32-byte encoding
    of the Scalar value.

>   - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a
>     32-byte string. This function can fail if the input does not represent a Scalar

Reviewer 4: Serialization and deserialization must be symmetrical:

```suggestion
  - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a
    little-endian 32-byte string. This function can fail if the input does not represent a Scalar
```

'''''''''''''''''''''''''''''

This ciphersuite uses Pallas for the Group and BLAKE2b-512 for the Hash function `H`
meant to produce signatures indistinguishable from RedPallas Sapling Spend
Authorization Signatures as specified in [#protocol-concretespendauthsig]_.

- Group: Pallas [#protocol-pallasandvesta]_

  - Order: 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001 (see [#protocol-pallasandvesta]_)
  - Identity: as defined in [#protocol-pallasandvesta]_
  - RandomScalar: Implemented by generating a random 64-byte string and invoking
    DeserializeScalar on the result
  - RandomNonZeroScalar: Implemented by generating a random 32-byte string that
    is not equal to the all-zero string and invoking DeserializeScalar on the result.
  - SerializeElement(P): Implemented as :math:`\mathsf{repr}_\mathbb{P}(P)` as defined in [#protocol-pallasandvesta]_
  - DeserializeElement(P): Implemented as :math:`\mathsf{abst}_\mathbb{P}(P)` as defined in [#protocol-pallasandvesta]_
  - SerializeScalar: Implemented by outputting the little-endian 32-byte encoding
    of the Scalar value.
  - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a

>     32-byte string. This function can fail if the input does not represent a Scalar

Reviewer 4: ```suggestion
    little-endian 32-byte string. This function can fail if the input does not represent a Scalar
```

To add re-randomization to FROST, follow the specification [#FROST]_ with the
following modifications.

Randomizer Generation
'''''''''''''''''''''

A new helper function is defined, which computes RedDSA.GenRandom:

::

  randomizer_generate():

  Inputs:
  - None

  Outputs: randomizer, a Scalar

  def randomizer_generate():
    randomizer_input = random_bytes(64)
    return H2(randomizer_input)

Binding Factor Computation
''''''''''''''''''''''''''

The `compute_binding_factor` function is changed to receive the `randomizer`
as follows: ::

  Inputs:
  - encoded_commitment_list, an encoded commitment list (as computed
    by encode_group_commitment_list)
  - msg, the message to be signed.
  - randomizer, the randomizer Scalar.

  Outputs: A Scalar representing the binding factor

  def compute_binding_factor(encoded_commitment_list, msg, randomizer):
    msg_hash = H3(msg)
    rho_input = encoded_commitment_list || msg_hash || G.SerializeScalar(randomizer)
    binding_factor = H1(rho_input)
    return binding_factor

Round Two - Signature Share Generation
''''''''''''''''''''''''''''''''''''''

In Round Two, the Coordinator generates a random scalar `randomizer` by calling

> `randomizer_generate` and sends it to each signer, over a secret channel,

Author: ```suggestion
`randomizer_generate` and sends it to each signer, over a confidential and authenticated channel,
```

along with the message and the set of signing commitments. (Note that this differs
from regular FROST which just requires an authenticated channel.)

The `sign` function is changed to receive `randomizer` and incorporate it into the
computation of the binding factor. It is specified as the following: ::

  Inputs:
  - identifier, Identifier i of the signer.

>     Note: identifier will never equal 0.

Author: ```suggestion
    Note: FROST spec requires that identifier will never equal 0.
```
(clarify? maybe also change the RFC?)

  - sk_i, Signer secret key share, a Scalar.
  - group_public_key, public key corresponding to the group signing key,
    an Element in G.
  - nonce_i, pair of Scalar values (hiding_nonce, binding_nonce) generated in

>     round one.

Author: Write an overview of round1 above

Author: done in 80a22f4

  - msg, the message to be signed (sent by the Coordinator).
  - commitment_list =
      [(j, hiding_nonce_commitment_j, binding_nonce_commitment_j), ...], a
    list of commitments issued in Round 1 by each signer and sent by the Coordinator.
    Each element in the list indicates the signer identifier j and their two commitment

>     Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).s

Author: ```suggestion
    Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).
```

(Check RFC too)

    # Compute the per-message challenge
    challenge = compute_challenge(group_commitment, randomized_group_public_key, msg)

    # Compute the signature share
    (hiding_nonce, binding_nonce) = nonce_i
    sig_share = hiding_nonce + (binding_nonce * binding_factor) + (lambda_i * sk_i * challenge)

    return sig_share

Signature Share Verification and Aggregation
''''''''''''''''''''''''''''''''''''''''''''

The `verify_signature_share` is changed to incorporate the randomizer,
as follows: ::

  Inputs:
  - identifier, Identifier i of the signer. Note identifier will never equal 0.
  - PK_i, the public key for the ith signer, where PK_i = G.ScalarBaseMult(sk_i),
    an Element in G
  - comm_i, pair of Element values in G (hiding_nonce_commitment, binding_nonce_commitment)
    generated in round one from the ith signer.
  - sig_share_i, a Scalar value indicating the signature share as produced in
    round two from the ith signer.
  - commitment_list =
      [(j, hiding_nonce_commitment_j, binding_nonce_commitment_j), ...], a
    list of commitments issued in Round 1 by each signer, where each element
    in the list indicates the signer identifier j and their two commitment

>     Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).s

Author: ```suggestion
    Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).
```

(Check RFC)

    # Compute the binding factor
    binding_factor = compute_binding_factor(encoded_commitments, msg, randomizer)

    # Compute the group commitment
    group_commitment = compute_group_commitment(commitment_list, binding_factor)

    # Compute the commitment share
    (hiding_nonce_commitment, binding_nonce_commitment) = comm_i
    comm_share = hiding_nonce_commitment + (binding_nonce_commitment * binding_factor)

    # Compute the challenge
    challenge = compute_challenge(group_commitment, randomized_group_public_key, msg)

    # Compute Lagrange coefficient
    participant_list = participants_from_commitment_list(commitment_list)
    lambda_i = derive_lagrange_coefficient(identifier, participant_list)

    # Compute relation values
    l = G.ScalarBaseMult(sig_share_i)
    r = comm_share + ((challenge * lambda_i) * PK_i)

    return l == r

The `aggregate` function is changed to incorporate the randomizer as follows: ::

  Inputs:
  - group_commitment, the group commitment returned by compute_group_commitment,
    an Element in G.
  - sig_shares, a set of signature shares z_i, Scalar values, for each signer,

>     of length NUM_SIGNERS, where MIN_SIGNERS <= NUM_SIGNERS <= MAX_SIGNERS.

Author: Mention that it's the same set from round 1 (also check RFC)

Author: Done in 80a22f4

  - challenge, the challenge returned by compute_challenge, a Scalar.
  - randomizer, the randomizer Scalar.

  Outputs:
  - (R, z), a Schnorr signature consisting of an Element R and Scalar z.
  - randomized_group_public_key, the randomized group public key

  def aggregate(group_commitment, sig_shares, group_public_key, challenge, randomizer):
    randomized_group_public_key = group_public_key + G * randomizer
    z = 0
    for z_i in sig_shares:
      z = z + z_i
    return (group_commitment, z + randomizer * challenge), randomized_group_public_key

Ciphersuites
------------

FROST(Jubjub, BLAKE2b-512)
'''''''''''''''''''''''''''''

This ciphersuite uses Jubjub for the Group and BLAKE2b-512 for the Hash function `H`
meant to produce signatures indistinguishable from RedJubjub Sapling Spend
Authorization Signatures as specified in [#protocol-concretespendauthsig]_.

- Group: Jubjub [#protocol-jubjub]_

  - Order: 6554484396890773809930967563523245729705921265872317281365359162392183254199 (see [#protocol-jubjub]_)
  - Identity: as defined in [#protocol-jubjub]_
  - RandomScalar: Implemented by generating a random 64-byte string and invoking

>     DeserializeScalar on the result

Author: Note: there's a bug in this that will be fixed after porting to RFC v7 (need to loop until a valid value is found)

Author: Change to: generate an integer uniformly distributed in the range [0, 2^512-1] and reduce modulo G.Order

(Also check RFC)

Author: Consider not describing how to do it, mention it as a non-normative note
(Also check RFC)

Author: V7 specifies this as, for example:

```
RandomScalar: Implemented by repeatedly generating a random 32-byte string and invoking DeserializeScalar on the result until success.
```

i.e. it uses rejection sampling instead of "wide reduction". I'm keen on keeping it that way for consistency with the RFC, but I'll discuss changing the RFC.

  - RandomNonZeroScalar: Implemented by generating a random 32-byte string that

>     is not equal to the all-zero string and invoking DeserializeScalar on the result.

Author: Note: this is not used in the spec and will be removed after porting to v7

  - SerializeElement(P): Implemented as :math:`\mathsf{repr}_\mathbb{J}(P)` as defined in [#protocol-jubjub]_

>   - DeserializeElement(P): Implemented as :math:`\mathsf{abst}_\mathbb{J}(P)` as defined in [#protocol-jubjub]_

Author: Add note that it's fallible

Author: Done in 80a22f4

  - SerializeScalar: Implemented by outputting the little-endian 32-byte encoding
    of the Scalar value.
  - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a
    32-byte string. This function can fail if the input does not represent a Scalar
    between the value 0 and G.Order() - 1.

- Hash (`H`): BLAKE2b-512 [#BLAKE]_ (BLAKE2b with 512-bit output and 16-byte personalization string),
  and Nh = 64.

  - H1(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubR", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.
  - H2(m): Implemented by computing BLAKE2b-512("Zcash_RedJubjubH", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.
    (This is equivalent to :math:`\mathsf{H}^\circledast(m)`, as defined in
    [#protocol-concretereddsa]_ parametrized with [#protocol-jubjub]_.)

>   - H3(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubDi", m)

Author: ```suggestion
  - H3(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubD", m)
```

  - H4(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubN", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.

FROST(Pallas, BLAKE2b-512)
'''''''''''''''''''''''''''''

> This ciphersuite uses Pallas for the Group and BLAKE2b-512 for the Hash function `H`
> meant to produce signatures indistinguishable from RedPallas Sapling Spend
> Authorization Signatures as specified in [#protocol-concretespendauthsig]_.
> 

Reviewer: ```suggestion
This ciphersuite uses Pallas for the Group and BLAKE2b-512 for the Hash function `H`
meant to produce signatures indistinguishable from RedPallas Orchard Spend
Authorization Signatures as specified in [#protocol-concretespendauthsig]_.

```

- Group: Pallas [#protocol-pallasandvesta]_

  - Order: 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001 (see [#protocol-pallasandvesta]_)
  - Identity: as defined in [#protocol-pallasandvesta]_
  - RandomScalar: Implemented by generating a random 64-byte string and invoking
    DeserializeScalar on the result
  - RandomNonZeroScalar: Implemented by generating a random 32-byte string that
    is not equal to the all-zero string and invoking DeserializeScalar on the result.
  - SerializeElement(P): Implemented as :math:`\mathsf{repr}_\mathbb{P}(P)` as defined in [#protocol-pallasandvesta]_
  - DeserializeElement(P): Implemented as :math:`\mathsf{abst}_\mathbb{P}(P)` as defined in [#protocol-pallasandvesta]_
  - SerializeScalar: Implemented by outputting the little-endian 32-byte encoding
    of the Scalar value.
  - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a

>     32-byte string. This function can fail if the input does not represent a Scalar

Author: Same comments as above

>     between the value 0 and G.Order() - 1.

Author: Use [0, G.Order()-1] (also above)

Author: Check function notation

Author: Done

Author: Range notation fixed in 80a22f4.

Regarding function notation, the RFC uses `Order()` when it's called; I'll check if we should add `()` here in the  ciphersuites definitions for consistency

When employing re-randomized FROST as specified in this ZIP, the goal is to
split the spend authorization private key among multiple participants. This means
that the proof generation will still be generated be a single participant, likely
the one that created the transaction in the first place and that will actually
submit the transaction to the network. Note that his user already controls the
privacy of the transaction since they are responsible for creating the proof.

This fits well into the "Coordinator" role from the FROST specification. The
Coordinator is responsible for sending the message to be signed to all participants,
and to aggregate the signature shares.

With those considerations in mind, the threat model considered in this ZIP is:

- The Coordinator is trusted with the privacy of the transaction (which includes
  the unlinkability property). A rogue Coordinator will be able to break
  unlinkability and privacy, but should not be able to create signed transactions
  without the approval of `MIN_SIGNERS` participants, as specified in FROST.

Non-requirements
================

- The control of the transaction privacy is not distributed among participants

>   and is trusted with the Coordinator.

Reviewer 3: I don't believe this is effectively worded given the participants learn of the randomizer. Accordingly, every participant is also trusted not to break unlinkability. Re-defining sign to take in the randomized key is possible, removing this possibility.

Author: Good point, fixed in 3a96a2f

    challenge = compute_challenge(group_commitment, randomized_group_public_key, msg)

    # Compute Lagrange coefficient
    participant_list = participants_from_commitment_list(commitment_list)
    lambda_i = derive_lagrange_coefficient(identifier, participant_list)

    # Compute relation values
    l = G.ScalarBaseMult(sig_share_i)
    r = comm_share + ((challenge * lambda_i) * PK_i)

    return l == r

The `aggregate` function is changed to incorporate the randomizer as follows: ::

  Inputs:
  - group_commitment, the group commitment returned by compute_group_commitment,
    an Element in G.
  - sig_shares, a set of signature shares z_i, Scalar values, for each signer,
    of length NUM_SIGNERS, where MIN_SIGNERS <= NUM_SIGNERS <= MAX_SIGNERS.
  - group_public_key, public key corresponding to the group signing key,
  - challenge, the challenge returned by compute_challenge, a Scalar.
  - randomizer, the randomizer Scalar.

  Outputs:
  - (R, z), a Schnorr signature consisting of an Element R and Scalar z.
  - randomized_group_public_key, the randomized group public key

  def aggregate(group_commitment, sig_shares, group_public_key, challenge, randomizer):

>     randomized_group_public_key = randomizer * group_public_key

Reviewer 3: randomizer * G + group_public_key

Author: Fixed in 54dbd57d768eadd9d7ee44d79869fc1d3f691f67

shielded transaction:

- The transaction is created.
- The transaction is signed with a re-randomized version of the user's spend
  authorization private key.
- The zero-knowledge proof for the transaction is created with the randomizer
  as an auxiliary (secret) input, among others.

When employing re-randomized FROST as specified in this ZIP, the goal is to
split the spend authorization private key :math:`\mathsf{ask}` among multiple
possible signers. This means that the proof generation will still be generated
by a single participant, likely the one that created the transaction in the first
place. Note that this user already controls the privacy of the transaction since
they are responsible for creating the proof.

This fits well into the "Coordinator" role from the FROST specification [#frost-protocol]_.
The Coordinator is responsible for sending the message to be signed to all participants,
and to aggregate the signature shares.

With those considerations in mind, the threat model considered in this ZIP is:

- The Coordinator is trusted with the privacy of the transaction (which includes
  the unlinkability property). A rogue Coordinator will be able to break

>   unlinkability and privacy, but should not be able to create signed transactions

Reviewer 3: Can I re-iterate all participants are able to break the unlinkability of the TX, not just the Coordinator? The discrete log of the re-randomization is sufficient as a proof.

Author: This item is only about the Coordinator, the next item talks about key share holders. But I made it more explicit in 1342ead

Reviewer 3: Ah, sorry, missed that. My bad there. Thanks for covering ti :+1:

  - SerializeScalar: Implemented by outputting the little-endian 32-byte encoding
    of the Scalar value.
  - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a
    little-endian 32-byte string. This function can fail if the input does not represent a Scalar
    in the range \[0, G.Order() - 1\].

- Hash (`H`): BLAKE2b-512 [#BLAKE]_ (BLAKE2b with 512-bit output and 16-byte personalization string),
  and Nh = 64.

  - H1(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubR", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.
  - H2(m): Implemented by computing BLAKE2b-512("Zcash_RedJubjubH", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.
    (This is equivalent to :math:`\mathsf{H}^\circledast(m)`, as defined in
    [#protocol-concretereddsa]_ parametrized with [#protocol-jubjub]_.)
  - H3(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubD", m)

>   - H4(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubN", m), interpreting

Reviewer: TODO: Update to include H5

::

>   ZIP: Unassigned {numbers are assigned by ZIP editors}

Reviewer 5: ```suggestion
  ZIP: 312
```

  Title: FROST for Spend Authorization Signatures
  Owners: Conrado Gouvea <conrado@zfnd.org>
          Chelsea Komlo <ckomlo@uwaterloo.ca>
          Deirdre Connolly <deirdre@zfnd.org>
  Credits: First Credited

>            ...

Reviewer 5: If there are no credited people (excluding Owners), delete these lines. If there are, just give names, not email addresses.

Author: Done in https://github.com/zcash/zips/pull/662

  Status: Draft
  Category: Wallet
  Created: 2022-08-dd

>   License: {usually MIT}

Reviewer 5: Please specify the license (e.g. `  License: MIT`).

Author: Done in https://github.com/zcash/zips/pull/662

>   Pull-Request: <https://github.com/zcash/zips/pull/TODO>

Reviewer 5: ```suggestion
  Discussions-To: <https://github.com/zcash/zips/issues/382>
  Pull-Request: <https://github.com/ZcashFoundation/zips/pull/3>
```

Author: Done in https://github.com/zcash/zips/pull/662

Terminology
===========

{Edit this to reflect the key words that are actually used.}
The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to
be interpreted as described in RFC 2119. [#RFC2119]_

The terms below are to be interpreted as follows:

Unlinkability

>   The impossibility of determining if two transactions were generated by the same

Reviewer 5: ```suggestion
  The property of statistical independence of signatures from the signers' long-term keys, ensuring that
  (for perfectly uniform generation of Randomizers and no leakage of metadata) it is impossible to determine whether two transactions were generated by the same
```
